src/ast.rs

```diff
@@ -0,0 +1,1617 @@
+        parent: Option<Id<Node>>,
+        start: usize,
+        end: usize,
+        raw: Wtf16,
+        min: Id<Node>,
+        max: Id<Node>,
+    ) -> Self {
+        Self::CharacterClassRange(CharacterClassRange {
+            _base: NodeBase {
+                _arena_id: Default::default(),
+                parent,
+                start,
+                end,
+                raw,
+            },
+            min,
+            max,
+        })
+    }
+
+    pub fn new_class_intersection(
+        parent: Option<Id<Node>>,
+        start: usize,
+        end: usize,
+        raw: Wtf16,
+        left: Id<Node>,
+        right: Id<Node>,
+    ) -> Self {
+        Self::ClassIntersection(ClassIntersection {
+            _base: NodeBase {
+                _arena_id: Default::default(),
+                parent,
+                start,
+                end,
+                raw,
+            },
+            left,
+            right,
+        })
+    }
+
+    pub fn new_class_subtraction(
+        parent: Option<Id<Node>>,
+        start: usize,
+        end: usize,
+        raw: Wtf16,
+        left: Id<Node>,
+        right: Id<Node>,
+    ) -> Self {
+        Self::ClassSubtraction(ClassSubtraction {
+            _base: NodeBase {
+                _arena_id: Default::default(),
+                parent,
+                start,
+                end,
+                raw,
+            },
+            left,
+            right,
+        })
+    }
+
+    pub fn new_class_string_disjunction(
+        parent: Option<Id<Node>>,
+        start: usize,
+        end: usize,
+        raw: Wtf16,
+        alternatives: Vec<Id<Node>>,
+    ) -> Self {
+        Self::ClassStringDisjunction(ClassStringDisjunction {
+            _base: NodeBase {
+                _arena_id: Default::default(),
+                parent,
+                start,
+                end,
+                raw,
+            },
+            alternatives,
+        })
+    }
+
+    pub fn new_string_alternative(
+        parent: Option<Id<Node>>,
+        start: usize,
+        end: usize,
+        raw: Wtf16,
+        elements: Vec<Id<Node>>,
+    ) -> Self {
+        Self::StringAlternative(StringAlternative {
+            _base: NodeBase {
+                _arena_id: Default::default(),
+                parent,
+                start,
+                end,
+                raw,
+            },
+            elements,
+        })
+    }
+
+    pub fn new_reg_exp_literal(
+        parent: Option<Id<Node>>,
+        start: usize,
+        end: usize,
+        raw: Wtf16,
+        pattern: Id<Node>,
+        flags: Id<Node>,
+    ) -> Self {
+        Self::RegExpLiteral(RegExpLiteral {
+            _base: NodeBase {
+                _arena_id: Default::default(),
+                parent,
+                start,
+                end,
+                raw,
+            },
+            pattern,
+            flags,
+        })
+    }
+
+    pub fn as_backreference(&self) -> &Backreference {
```

Comment:
Might be interesting to have a macro that generates _all_ of these (for a given enum)?